File: Sources/Gamified/ViewModels/GameProfileViewModel.swift
//
//  GameProfileViewModel.swift
//  
//
//  Created by Nathan Fallet on 03/04/2023.
//

import Foundation

class GameProfileViewModel: ObservableObject {
    
    // Configuration
    
    let service: GamifiedService
    let username: String?
    let url: URL?
    
    @Published var experience: Experience
    @Published var achievements: [Achievement]
    
    // Initializer
    
    init(
        service: GamifiedService
    ) {
        self.service = service
        self.username = nil
        self.url = nil
        self.experience = Experience(total: 0)
        self.achievements = []
        DispatchQueue.global().async {
            let experienceData = service.getExperience()
            DispatchQueue.main.async {
                self.experience = experienceData
            }
            let achievementsData = service.getAchievements()
            DispatchQueue.main.async {
                self.achievements = achievementsData
            }
        }
    }
    
}

File: Gamified/src/main/java/me/nathanfallet/gamified/viewmodels/GameProfileViewModel.kt
package me.nathanfallet.gamified.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import me.nathanfallet.gamified.models.Achievement
import me.nathanfallet.gamified.models.Experience
import me.nathanfallet.gamified.services.GamifiedService

class GameProfileViewModel(
    application: Application
) : AndroidViewModel(application) {

    // Configuration

    val username: String? = null
    val url: String? = null

    private val _experience = MutableLiveData<Experience>()
    val experience: LiveData<Experience> = _experience

    private val _achievements = MutableLiveData<List<Achievement>>()
    val achievements: LiveData<List<Achievement>> = _achievements

    // Initializer

    init {
        viewModelScope.launch {
            _experience.value = GamifiedService.shared.getExperience()
            _achievements.value = GamifiedService.shared.getAchievements()
        }
    }

}